sdk/vaa/structs.go

```diff
@@ -367,11 +408,161 @@ func Unmarshal(data []byte) (*VAA, error) {
+	}
+	v.Hashes = make([]common.Hash, int(lenHashes))
+	for i := 0; i < int(lenHashes); i++ {
+		hash := [32]byte{}
+		if n, err := reader.Read(hash[:]); err != nil || n != 32 {
+			return nil, fmt.Errorf("failed to read hash [%d]: %w", i, err)
+		}
+		v.Hashes[i] = common.BytesToHash(hash[:])
+	}
+
+	lenObservations, err := reader.ReadByte()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read observations length: %w", err)
+	}
+
+	v.Observations = make([]*Observation, int(lenObservations))
+	for i := 0; i < int(lenObservations); i++ {
+		obsvIndex, err := reader.ReadByte()
+		if err != nil {
+			return nil, fmt.Errorf("failed to read Observation index [%d]: %w", i, err)
+		}
+
+		obsvLength := uint32(0)
+		if err := binary.Read(reader, binary.BigEndian, &obsvLength); err != nil {
+			return nil, fmt.Errorf("failed to read Observation length: %w", err)
+		}
+
+		obs := make([]byte, int(obsvLength))
+		if n, err := reader.Read(obs[:]); err != nil || n == 0 {
+			return nil, fmt.Errorf("failed to read Observation bytes [%d]: %w", n, err)
+		}
+
+		msgPub, err := UnmarshalBatchObservation(obs)
+		if err != nil {
+			return nil, fmt.Errorf("failed to unmarshal Observation VAA. %w", err)
+		}
+
+		v.Observations[i] = &Observation{
+			Index:       uint8(obsvIndex),
+			Observation: msgPub,
+		}
+	}
+
+	return v, nil
+}
+
+// Unmarshal deserializes the binary representation of an Observation (VAA)
+// within a BatchVAA. Different from VAA unmarshaling - no header fields.
+func UnmarshalBatchObservation(data []byte) (*VAA, error) {
+	if len(data) < minHeadlessVAALength {
+		return nil, fmt.Errorf("BatchVAA.Observation is too short")
+	}
+	v := &VAA{}
+
+	reader := bytes.NewReader(data[:])
+
+	unixSeconds := uint32(0)
+	if err := binary.Read(reader, binary.BigEndian, &unixSeconds); err != nil {
+		return nil, fmt.Errorf("failed to read timestamp: %w", err)
+	}
+	v.Timestamp = time.Unix(int64(unixSeconds), 0)
+
+	if err := binary.Read(reader, binary.BigEndian, &v.Nonce); err != nil {
+		return nil, fmt.Errorf("failed to read nonce: %w", err)
+	}
+
+	if err := binary.Read(reader, binary.BigEndian, &v.EmitterChain); err != nil {
+		return nil, fmt.Errorf("failed to read emitter chain: %w", err)
+	}
+
+	emitterAddress := Address{}
+	if n, err := reader.Read(emitterAddress[:]); err != nil || n != 32 {
+		return nil, fmt.Errorf("failed to read emitter address [%d]: %w", n, err)
+	}
+	v.EmitterAddress = emitterAddress
+
+	if err := binary.Read(reader, binary.BigEndian, &v.Sequence); err != nil {
+		return nil, fmt.Errorf("failed to read sequence: %w", err)
+	}
+
+	if err := binary.Read(reader, binary.BigEndian, &v.ConsistencyLevel); err != nil {
+		return nil, fmt.Errorf("failed to read commitment: %w", err)
+	}
+
+	payload := make([]byte, InternalTruncatedPayloadSafetyLimit)
+	n, err := reader.Read(payload)
+	if err != nil || n == 0 {
+		return nil, fmt.Errorf("failed to read payload [%d]: %w", n, err)
+	}
+
+	v.Payload = payload[:n]
```

Comment:
Yes! Fixed.

I aired toward duplicating rather than refactoring because I was worried about making too big of a change and not having anyone to review. I'm very happy to do it properly.